### PR TITLE
Improved log1p implementation for complex inputs

### PR DESCRIPTION
**Pull request by tringwald**
This PR improves the implementation of `torch.log1p` for complex inputs as mentioned in issue #107022. The new implementation is based on the insights provided in https://github.com/numpy/numpy/pull/22611#issuecomment-1667945354.

cc @lezcano

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
Since this function is also in `c10_complex_math`, you don't need to specify the namespace, but fair enough.

Thank you for the quick fix! Feel free to merge the PR yourself via @pytorchbot

**Comment by lezcano** [comment]
So, as discussed in https://github.com/numpy/numpy/pull/22611#issuecomment-1667945354, the implementation in mac seems to be failing. The relative error is small (6e-6) but higher than the tolerance of 1e-6.
Given that this operation is meant to be used as a more precise version of `log`, I think we should fix it. I propose you resort to the previous implementation (which was passing the tests) for mac via compile flags.

**Comment by lezcano** [comment]
And could you also leave a comment as to why we're doing that.

**Comment by tringwald** [comment]
I added a fallback to the old version in case `__APPLE__` or `__MACOSX` is defined. I also removed the namespace specifier for c10_complex_math::log.

**Comment by lezcano** [comment]
The failures seem real (although CI is not particularly helpful). Could you look into those?

**Comment by tringwald** [comment]
There seems to be a problem with building `LogcumsumexpKernel.cu`. I'll have a look.

**Comment by tringwald** [comment]
So far, I've not been able to reproduce the build errors on my local setup. I've checked the failing `linux-jammy` and `linux-bionic` builds with both `gcc` and `clang` inside of Docker.
Given the `Killed` output in all of the failing builds, is the build server maybe running out of memory?

File: c10/util/complex_math.h [hunk 1]
```diff
@@ -316,6 +321,17 @@ C10_HOST_DEVICE inline c10::complex<T> log1p(const c10::complex<T>& z) {
...
   }
+#else
+  // Based on https://github.com/numpy/numpy/pull/22611#issuecomment-1667945354
+  c10::complex<T> u = z + T(1);
+  if (u == T(1)) {
+    return z;
+  } else if (u - T(1) == z) {
+    return log(u);
+  } else {
+    return log(u) * (z / (u - T(1)));
```
Comment (peterbell10) [inline c10/util/complex_math.h:332]: Sadly these complex functions are far more costly than the real counterparts and it looks like the build is running out of memory when compiling them. A shot in the dark, there might be a compile time improvement by only calling `log(u)` once:
```cpp
auto log_u = log(u);
if (u - T(1) != z) {
    return log_u * (z / (u - T(1));
}
return log_u;
```

**Comment by lezcano** [comment]
All those failures seem unrelated. Can you rebase off `viable/strict` (it's a branch that should have CI green... albeit sometimes it doesn't, but yeah) to clean all these failures?

**Comment by lezcano** [comment]
Há! That did it! nvcc is sometimes a bit thick...

**Pull request closed without merging**